lib/chibi/math/prime.scm

```diff
@@ -175,57 +189,109 @@
          ((prime? n) n)
          (else (lp (+ n 2)))))))))
 
+;; Given an initial value \var{r1} representing the (empty)
+;; factorization of 1 and a procedure \var{put}
+;; (called as \scheme{(\var{put} \var{r} \var{p} \var{k})})
+;; that, given prior representation \var{r},
+;; adds a prime factor \var{p} of multiplicity \var{k},
+;; returns a factorization function which returns the factorization
+;; of its non-zero integer argument \var{n} in this representation.
+;; The following optional arguments, if provided, specialize \var{put}
+;; for particular primes:
+;; \var{put2:}  for \var{p}=2,  called as \scheme{(\var{put2} \var{r} \var{k})})
+;; \var{put-1:} for \var{p}=-1, called as \scheme{(\var{put-1} \var{r})}).
+(define (make-factorizer r1 put . o)
+  (let-keywords o ((put2  (lambda (r k) (put r  2 k)))
+		   (put-1 (lambda (r)   (put r -1 1))))
```

Comment:
Keywords are overkill, you can just use let-optionals.  Actually, I should add define-opt* so you can

```
(define-opt* (make-factorizer r1 put (put2 (lambda (r k) (put r 2 k))) (put-1 (lambda (r) (put r -1 1))))
  ...)
```

And actually, since you only call this in 3 places and put2 and put-1 are different every time, you could also just make them required arguments and simplify things.

Reply:
> put2 and put-1 are different every time
no, all of the `put`s are the same for `factor-alist` and the (commented out) short version of `factor` (i.e., apart from taking different arguments, which means having them both required would be painful).  

I used keywords because I wasn't sure what else `make-factorizer` might eventually be used for (was pondering if `for-each-factor` is the Right Thing).  Not wedded to them.

Reply:
ok, it's using `let-optionals` now.

> I should add define-opt*

I just looked at SRFI 89 and SRFI 177 and am sensing a huge rabbit hole here.

Reply:
Yes, that is indeed a rabbit hole.  It has some hopes of getting standardized soon though.

Reply:
> I will try to find a false positive.

 and whump (see latest commit).   It turns out any [Carmichael number](https://mathworld.wolfram.com/CarmichaelNumber.html) with smallest prime factor ≥ 59 will do, anything of the form (6k+1)(12k+1)(18k+1) where all 3 of those factors are prime is a Carmichael number, and there are [scads of these](https://oeis.org/A046025).

Reply:
It also occurs to me that sabotaging the random number generator is perhaps overkill.  

I.e., if the fact that  
 `(test #t (miller-rabin-composite? 118901521))` 
 fails 97% of the time on the old code even if we leave the random number generator alone is sufficient for your purposes (because the RNG only gets 4 tries to find a witness outside of 2..53 which means hitting something that's a multiple of one of those 3 prime factors (smallest is 241, so that's a 1/241 chance for that one; the others are harder),
 
 then we can reduce the test to 1 line and punt on the thread-safety issues inherent in bashing `random-integer` (i.e., if you care about ever trying to run tests concurrently on multiple cores or something ... buh)

Reply:
> we can reduce the test to that 1 line
doing this, especially now that I've figured out[ this is all deterministic anyway (see other thread)](https://github.com/ashinn/chibi-scheme/pull/752#discussion_r660947368).